cmd/gateway-main.go

```diff
@@ -45,6 +45,57 @@ var (
 	}
 )
 
+// GatewayLocker implements custom NeNSLock implementation
+type GatewayLocker struct {
+	ObjectLayer
+	nsMutex *nsLockMap
+}
+
+// NewNSLock - implements gateway level locker
+func (l *GatewayLocker) NewNSLock(ctx context.Context, bucket string, objects ...string) RWLocker {
+	return l.nsMutex.NewNSLock(ctx, nil, bucket, objects...)
+}
+
+// Walk - implements common gateway level Walker, to walk on all objects recursively at a prefix
+func (l *GatewayLocker) Walk(ctx context.Context, bucket, prefix string, results chan<- ObjectInfo) error {
+	walk := func(ctx context.Context, bucket, prefix string, results chan<- ObjectInfo) error {
+		var marker string
+		for {
+			// set maxKeys to '0' to list maximum possible objects in single call.
+			loi, err := l.ObjectLayer.ListObjects(ctx, bucket, prefix, marker, "", 0)
```

Comment:
when maxKeys is set to 0, s3 returns an empty list even if bucket has objects. maxKeys should be set to 1000 unless it is NAS, imo

Reply:
Looks like a minio-go bug @poornas if maxKeys is set to 0 it should return 1000 keys. 

```
        // Set max keys.
        if maxkeys > 0 {
                urlValues.Set("max-keys", fmt.Sprintf("%d", maxkeys))
        }
```